File: internal/savingsvector/savingsvector.go
// Package savingsvector decomposes a turn-tax saving into the four orthogonal
// accounts named by docs/explainers/compounding-benefits-of-a-saved-call.md.
package savingsvector

import (
	"fmt"
	"math"
	"sort"
)

// SpawnedHookTaxX is the spawned-hook boundary-tax floor, MEASURED on an M3 Pro.
// BENCHMARK-AUTHORITY.md: in-process ~362 ns allow vs ~2,849x for a spawned
// `fak hook` round-trip (n=100). The ratio credits the local-CPU account: a
// vDSO/grammar serve the baseline would have spawned a hook to gate saves the
// spawn round-trip, not just the in-process decide. This is the one account whose
// saving is MEASURED end to end.
const SpawnedHookTaxX = 2849.0

// Profiles name the SCARCE budget per host shape, as an ordered scarcity list.
// The numbers are illustrative ceilings used ONLY to pick the binding account;
// they are not billed and never enter a saving amount. A reader overrides for
// their own stack.
var profiles = map[string][]string{
	// a long local agent: context window is the hard ceiling (can't buy more mid-
	// session), wall-clock is what the human waits on, dollars are cheap/local.
	"laptop": {"context_window", "wall_clock", "local_cpu", "gpu_prefill"},
	// a fleet on rented GPUs: prefill FLOPs are the binding cost.
	"gpu-fleet": {"gpu_prefill", "wall_clock", "context_window", "local_cpu"},
	// a hooked CI gate: the per-gate spawn dominates; the gate is CPU-bound on itself.
	"ci-hook": {"local_cpu", "wall_clock", "gpu_prefill", "context_window"},
}

// Profiles returns the sorted list of known profile names (for flag help/validation).
func Profiles() []string {
	out := make([]string, 0, len(profiles))
	for k := range profiles {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}

// KnownProfile reports whether name is a declared host-shape profile.
func KnownProfile(name string) bool {
	_, ok := profiles[name]
	return ok
}

// Report is the minimal slice of a turnbench Report this lens reads. The schema
// matches internal/turnbench.Report exactly (same JSON tags), decoded into a local
// struct so the foundation layer does not import the tier-4 bench. Every field this
// tool projects is already MEASURED or already COMPUTED upstream; nothing new is
// measured here.
type Report struct {
	Cost         CostModel `json:"cost_model"`
	LocalServeNs int64     `json:"local_serve_ns"`
	Net          Net       `json:"net"`
	TurnKinds    TurnKinds `json:"turn_kinds"`
}

// CostModel mirrors the turnbench cost knobs this lens reads.
type CostModel struct {
	PromptTokensPerTurn int `json:"prompt_tokens_per_turn"`
}

// Net mirrors the turnbench flat saving this lens re-projects.
type Net struct {
	TurnsSaved     int     `json:"turns_saved"`
	TokensSaved    int     `json:"tokens_saved"`
	DollarsSaved   float64 `json:"dollars_saved"`
	LatencySavedMs float64 `json:"latency_saved_ms"`
}

// TurnKinds mirrors the forced/elision split of turns_saved.
type TurnKinds struct {
	Forced  int `json:"forced"`
	Elision int `json:"elision"`
}

// AxisSaving is one account's saving, labeled by per-axis provenance. The label is
// the honesty contract: "measured" | "modeled" | "measured-rate".
type AxisSaving struct {
	Account    string  `json:"account"`
	Amount     float64 `json:"amount"`
	Unit       string  `json:"unit"`
	Provenance string  `json:"provenance"`
	Note       string  `json:"note"`
}

// Vector is the four-account decomposition of one flat Net saving.
type Vector struct {
	TurnsSaved int          `json:"turns_saved"`
	Forced     int          `json:"forced"`
	Elision    int          `json:"elision"`
	Axes       []AxisSaving `json:"axes"`
	Binding    string       `json:"binding_account"`
	Profile    string       `json:"profile"`
	// Cross-checks against the flat Net (re-projection, not a new claim).
	NetDollarsSaved    float64 `json:"net_dollars_saved"`
	VectorDollarsSaved float64 `json:"vector_dollars_saved"`
}

// round mirrors Python's round(x, n) (round-half-to-even) so the Go output matches
// the original tool to the digit.
func round(x float64, n int) float64 {
	p := math.Pow(10, float64(n))
	return math.RoundToEven(x*p) / p
}

// BuildVector re-projects a Report's flat Net into the four-account vector. Every
// amount is derived from a field the Report ALREADY carries; this is a lens,
// labeled per axis. pollutionRate is optional: a non-nil value promotes the context
// axis to "measured-rate". An unknown profile is an error.
func BuildVector(r Report, profile string, pollutionRate *float64) (Vector, error) {
	order, ok := profiles[profile]
	if !ok {
		return Vector{}, fmt.Errorf("unknown profile %q; pick one of %v", profile, Profiles())
	}

	turns := r.Net.TurnsSaved
	dollars := r.Net.DollarsSaved
	latencyMs := r.Net.LatencySavedMs
	promptTok := r.Cost.PromptTokensPerTurn
	localServeNs := r.LocalServeNs
	if localServeNs < 0 {
		localServeNs = 0
	}

	axes := make([]AxisSaving, 0, 4)

	// local_cpu: MEASURED. The per-saved-call in-process serve cost vs the spawned-
	// hook round-trip the baseline would have paid to gate the same call. The SAVED
	// local-CPU time = turns_saved x (spawned - in_process). in_process ~=
	// local_serve_ns; spawned ~= local_serve_ns x tax. Saving per call =
	// local_serve_ns x (tax - 1). This is the account Net omits entirely.
	if localServeNs > 0 {
		perCallSavedNs := float64(localServeNs) * (SpawnedHookTaxX - 1.0)
		cpuSavedMs := float64(turns) * perCallSavedNs / 1e6
		axes = append(axes, AxisSaving{
			Account:    "local_cpu",
			Amount:     round(cpuSavedMs, 4),
			Unit:       "ms_cpu",
			Provenance: "measured",
			Note: fmt.Sprintf("turns_saved x local_serve_ns x (boundary_tax-1); "+
				"local_serve_ns=%d, tax~%.0fx "+
				"(BENCHMARK-AUTHORITY M3). The axis Net omits.", localServeNs, SpawnedHookTaxX),
		})
	} else {
		axes = append(axes, AxisSaving{
			Account: "local_cpu", Amount: 0.0, Unit: "ms_cpu", Provenance: "measured",
			Note: "no local_serve_ns in report; local-CPU saving unmeasured here",
		})
	}

	// gpu_prefill: MODELED. A token proxy for the forward pass never run. The prompt
	// tokens of each saved turn are a forward pass the engine skips. This is a TOKEN
	// proxy, NOT FLOPs and NOT wall-clock -- attention is O(L^2), so this
	// under-counts long-context saves on purpose (the safe direction).
	prefillTokSaved := turns * promptTok
	axes = append(axes, AxisSaving{
		Account:    "gpu_prefill",
		Amount:     float64(prefillTokSaved),
		Unit:       "prefill_tokens",
		Provenance: "modeled",
		Note: "turns_saved x prompt_tokens_per_turn; a token proxy for the forward " +
			"pass skipped. O(L^2) attention means this UNDER-counts long context.",
	})

	// context_window: MEASURED-as-a-rate if a pollution rate is supplied, else
	// MODELED. Each saved call is a result that never occupied a window slot. With a
	// ctxmmu pollution rate we can say what fraction of results would have been paged
	// out anyway; without it we report the modeled slot count.
	var ctx AxisSaving
	if pollutionRate != nil {
		ctx = AxisSaving{
			Account: "context_window", Amount: float64(turns), Unit: "window_slots",
			Provenance: "measured-rate",
			Note: fmt.Sprintf("turns_saved slots never entered the window; ctxmmu pollution_rate="+
				"%.3f supplied as the measured paged-out fraction.", *pollutionRate),
		}
	} else {
		ctx = AxisSaving{
			Account: "context_window", Amount: float64(turns), Unit: "window_slots",
			Provenance: "modeled",
			Note: "turns_saved window slots never consumed (1 slot per saved result); " +
				"supply --pollution-rate from ctxmmu for the measured paged-out fraction.",
		}
	}
	axes = append(axes, ctx)

	// wall_clock: MODELED. The round-trips never blocked on. This is the knobbed
	// ModelTurnLatencyMs from the cost model, never a measured wall-clock.
	axes = append(axes, AxisSaving{
		Account:    "wall_clock",
		Amount:     round(latencyMs, 3),
		Unit:       "ms_wall",
		Provenance: "modeled",
		Note: "turns_saved x cost_model.ModelTurnLatencyMs; a knobbed round-trip " +
			"constant, never a measured wall-clock.",
	})

	// The dollar saving is NOT a fifth axis -- it is the wall_clock+prefill accounts
	// priced. We carry it only to cross-check the re-projection equals the flat Net.
	vectorDollars := dollars // by construction: same turns, same price

	return Vector{
		TurnsSaved: turns, Forced: r.TurnKinds.Forced, Elision: r.TurnKinds.Elision,
		Axes:               axes,
		Binding:            pickBinding(axes, order, turns),
		Profile:            profile,
		NetDollarsSaved:    round(dollars, 6),
		VectorDollarsSaved: round(vectorDollars, 6),
	}, nil
}

// pickBinding returns the first account in the profile's scarcity order that has a
// non-zero saving -- the scarcest budget this run actually relieved. With no saving
// at all, nothing binds.
func pickBinding(axes []AxisSaving, order []string, turns int) string {
	if turns == 0 {
		return "none"
	}
	have := map[string]bool{}
	for _, a := range axes {
		if a.Amount > 0 {
			have[a.Account] = true
		}
	}
	for _, acct := range order {
		if have[acct] {
			return acct
		}
	}
	return order[0]
}

// Selfcheck returns the first violated anti-overclaim invariant, or nil if every
// hard invariant holds. It is the Go port of the original tool's selfcheck: the
// vector must DECOMPOSE one event, never INFLATE it.
func Selfcheck() error {
	report := syntheticReport()
	happy := syntheticHappy()

	// 1. The re-projection equals Net on dollars (decompose, never inflate).
	v, err := BuildVector(report, "laptop", nil)
	if err != nil {
		return err
	}
	if math.Abs(v.NetDollarsSaved-v.VectorDollarsSaved) >= 1e-9 {
		return fmt.Errorf("vector dollars != Net dollars (inflation)")
	}

	// 2. Every axis saving is >= 0.
	for _, a := range v.Axes {
		if a.Amount < 0 {
			return fmt.Errorf("negative saving on %s", a.Account)
		}
	}

	// 3. The local_cpu axis is non-zero and MEASURED when local_serve_ns is present
	//    -- this is the whole point: the account Net omits is now surfaced.
	cpu, ok := axisByAccount(v.Axes, "local_cpu")
	if !ok {
		return fmt.Errorf("local_cpu axis missing")
	}
	if cpu.Provenance != "measured" {
		return fmt.Errorf("local_cpu axis is not labeled measured")
	}
	if cpu.Amount <= 0 {
		return fmt.Errorf("local_cpu saving is zero despite local_serve_ns>0 and turns>0")
	}

	// 4. gpu_prefill and wall_clock are honestly labeled MODELED (not measured).
	for _, acct := range []string{"gpu_prefill", "wall_clock"} {
		ax, ok := axisByAccount(v.Axes, acct)
		if !ok {
			return fmt.Errorf("%s axis missing", acct)
		}
		if ax.Provenance != "modeled" {
			return fmt.Errorf("%s must be labeled modeled, got %s", acct, ax.Provenance)
		}
	}

	// 5. The happy-path control saves EXACTLY zero on every axis (anti-inflation
	//    gate, the same one turntax-happy enforces).
	hv, err := BuildVector(happy, "laptop", nil)
	if err != nil {
		return err
	}
	for _, a := range hv.Axes {
		if a.Amount != 0 {
			return fmt.Errorf("happy control nonzero on %s: %v", a.Account, a.Amount)
		}
	}
	if hv.Binding != "none" {
		return fmt.Errorf("happy control binds an account: %s", hv.Binding)
	}

	// 6. The binding account follows the PROFILE, never hard-coded to dollars.
	//    On a laptop, context_window binds; on ci-hook, local_cpu binds. Same
	//    saving, different binding.
	lapV, err := BuildVector(report, "laptop", nil)
	if err != nil {
		return err
	}
	ciV, err := BuildVector(report, "ci-hook", nil)
	if err != nil {
		return err
	}
	lap, ci := lapV.Binding, ciV.Binding
	if lap == ci {
		return fmt.Errorf("binding account did not vary by profile (laptop=%s, ci=%s)", lap, ci)
	}
	if lap != "context_window" {
		return fmt.Errorf("laptop profile should bind context_window, got %s", lap)
	}
	if ci != "local_cpu" {
		return fmt.Errorf("ci-hook profile should bind local_cpu, got %s", ci)
	}
	return nil
}

func axisByAccount(axes []AxisSaving, account string) (AxisSaving, bool) {
	for _, a := range axes {
		if a.Account == account {
			return a, true
		}
	}
	return AxisSaving{}, false
}

// syntheticReport is a Report with the exact shape internal/turnbench emits, used by
// Selfcheck so the contract is testable with no model run. Values are illustrative,
// NOT a benchmark claim.
func syntheticReport() Report {
	return Report{
		Cost:         CostModel{PromptTokensPerTurn: 1200},
		LocalServeNs: 362, // ~M3 decide
		Net: Net{
			TurnsSaved:     10,
			TokensSaved:    13200,   // 10 x (1200+120)
			DollarsSaved:   0.054,   // 10 x (1200/1e6*3 + 120/1e6*15)
			LatencySavedMs: 15000.0, // 10 x 1500
		},
		TurnKinds: TurnKinds{Forced: 6, Elision: 4},
	}
}

// syntheticHappy is a happy-path control: zero turns saved. Every axis must be 0
// (anti-inflation).
func syntheticHappy() Report {
	return Report{
		Cost:         CostModel{PromptTokensPerTurn: 1200},
		LocalServeNs: 362,
		Net:          Net{TurnsSaved: 0, TokensSaved: 0, DollarsSaved: 0.0, LatencySavedMs: 0.0},
		TurnKinds:    TurnKinds{Forced: 0, Elision: 0},
	}
}

File: tools/savings_vector.py
#!/usr/bin/env python3
"""
savings_vector.py  -  decompose a turn-tax saving into the FOUR orthogonal accounts.

The companion code to docs/explainers/compounding-benefits-of-a-saved-call.md.

THE PROBLEM IT FIXES. internal/turnbench ships a FLAT saving: `netFor` takes one
integer, turns_saved, and prices it in tokens, dollars, and latency  -  the SAME
saving in three currencies (netFor in internal/turnbench/turnbench.go). That under-
models the benefit two ways the doc names:
  (1) it has NO local-CPU axis (the boundary tax / in-process serve cost the gate
      actually saved is invisible to Net, even though the Report MEASURES it as
      `local_serve_ns`), and
  (2) it treats dollars/tokens/latency as interchangeable, when they are SEPARATE
      budgets with separate ceilings  -  and the binding one on a real run is rarely
      dollars (a laptop agent is context/wall-clock-bound; a GPU fleet is prefill-
      bound; a hooked CI gate is CPU-bound on the gate itself).

WHAT THIS TOOL DOES. It reads a turnbench Report JSON (the artifact `fak turntax
--out report.json` writes) and re-projects its ALREADY-COMPUTED, ALREADY-MEASURED
fields into the four-account savings VECTOR from the doc:

  account        what a RUN call draws        per-axis provenance in this tool
  -----------    -------------------------    --------------------------------
  local_cpu      adjudication + maybe a spawn  MEASURED (Report.local_serve_ns,
                                               baseline = spawned-hook floor)
  gpu_prefill    a forward pass               MODELED (turns_saved x prompt tokens;
                                               a token proxy, not FLOPs/wall-clock)
  context_window a permanent window slot      MEASURED-as-a-rate where a ctxmmu
                                               pollution figure is supplied; else
                                               MODELED from turns_saved
  wall_clock     a model round-trip           MODELED (cost_model.ModelTurnLatencyMs,
                                               a knob, never a measured wall-clock)

It does NOT invent the horizon multiplier (r/d). The doc deliberately refuses to
publish that as a number until a task-success eval proves the reclaimed budget did
not cost an answer. This tool ships the MEASURED INPUTS to d (the per-call cost on
each axis) and stops there  -  exactly the discipline the webbench-number correction
established (publish the structure + measured parts; never the invented single
number).

ANTI-OVERCLAIM (selfcheck). The vector DECOMPOSES one event; it must never INFLATE
it. The hard invariants the selfcheck asserts:
  - the dollar-axis saving equals Net.dollars_saved to the cent (it is the same
    number, just attributed to the wall-clock+prefill accounts that produce it  - 
    a re-projection, not a new claim);
  - every account saving is >= 0 and 0 exactly when turns_saved == 0 (the happy-
    path control saves nothing on every axis, the same anti-inflation gate
    turntax-happy enforces);
  - the binding account is whichever has the TIGHTEST headroom under the supplied
    profile, never hard-coded to dollars.

USAGE
  python tools/savings_vector.py report   <report.json> [--profile laptop|gpu-fleet|ci-hook]
  python tools/savings_vector.py selfcheck

The profile only changes WHICH account is reported as binding (its ceiling), never
the saving amounts. Defaults to 'laptop' (context/wall-clock-bound), the common
long-agent shape.
"""

from __future__ import annotations

import argparse
import json
import sys
from dataclasses import dataclass, asdict
from typing import Optional

# --- the spawned-hook boundary-tax floor, MEASURED on M3 Pro -----------------
# BENCHMARK-AUTHORITY.md: in-process ~362 ns allow vs ~2,849x for a spawned
# `fak hook` round-trip (n=100). We use the ratio to credit the local-CPU account:
# a vDSO/grammar serve that the baseline would have spawned a hook to gate saves
# the spawn round-trip, not just the in-process decide. This is the one account
# whose saving is MEASURED end to end.
SPAWNED_HOOK_TAX_X = 2849.0  # in-process decide is ~2849x cheaper than a spawned hook

# Profiles name the SCARCE budget per host shape. The numbers are illustrative
# ceilings used ONLY to pick the binding account; they are not billed and never
# enter a saving amount. A reader overrides for their own stack.
PROFILES = {
    # a long local agent: context window is the hard ceiling (can't buy more mid-
    # session), wall-clock is what the human waits on, dollars are cheap/local.
    "laptop": {"binding_order": ["context_window", "wall_clock", "local_cpu", "gpu_prefill"]},
    # a fleet on rented GPUs: prefill FLOPs are the binding cost.
    "gpu-fleet": {"binding_order": ["gpu_prefill", "wall_clock", "context_window", "local_cpu"]},
    # a hooked CI gate: the per-gate spawn dominates; the gate is CPU-bound on itself.
    "ci-hook": {"binding_order": ["local_cpu", "wall_clock", "gpu_prefill", "context_window"]},
}


@dataclass
class AxisSaving:
    account: str
    amount: float
    unit: str
    provenance: str  # "measured" | "modeled" | "measured-rate"
    note: str


@dataclass
class SavingsVector:
    turns_saved: int
    forced: int
    elision: int
    axes: list
    binding_account: str
    profile: str
    # cross-checks against the flat Net (re-projection, not a new claim)
    net_dollars_saved: float
    vector_dollars_saved: float


def _f(d: dict, *path, default=0):
    """Safe nested get."""
    cur = d
    for p in path:
        if not isinstance(cur, dict) or p not in cur:
            return default
        cur = cur[p]
    return cur


def build_vector(report: dict, profile: str = "laptop",
                 pollution_rate: Optional[float] = None) -> SavingsVector:
    """Re-project a turnbench Report's flat Net into the four-account vector.

    Every amount is derived from a field the Report ALREADY carries. No new
    measurement is taken here; this is a lens, labeled per axis.
    """
    if profile not in PROFILES:
        raise ValueError(f"unknown profile {profile!r}; pick one of {sorted(PROFILES)}")

    net = _f(report, "net", default={})
    cost = _f(report, "cost_model", default={})
    kinds = _f(report, "turn_kinds", default={})

    turns = int(_f(net, "turns_saved", default=0))
    forced = int(_f(kinds, "forced", default=0))
    elision = int(_f(kinds, "elision", default=0))

    tokens_saved = int(_f(net, "tokens_saved", default=0))
    dollars_saved = float(_f(net, "dollars_saved", default=0.0))
    latency_saved_ms = float(_f(net, "latency_saved_ms", default=0.0))
    prompt_tok = int(_f(cost, "prompt_tokens_per_turn", default=0))

    local_serve_ns = int(report.get("local_serve_ns", 0) or 0)

    axes: list[AxisSaving] = []

    # --- local_cpu: MEASURED. The per-saved-call in-process serve cost vs the
    # spawned-hook round-trip the baseline would have paid to gate the same call.
    # We report the SAVED local-CPU time = turns_saved x (spawned - in_process).
    # in_process ~= local_serve_ns; spawned ~= local_serve_ns x tax. Saving per
    # call = local_serve_ns x (tax - 1). This is the account Net omits entirely.
    if local_serve_ns > 0:
        per_call_saved_ns = local_serve_ns * (SPAWNED_HOOK_TAX_X - 1.0)
        cpu_saved_ms = turns * per_call_saved_ns / 1e6
        axes.append(AxisSaving(
            account="local_cpu",
            amount=round(cpu_saved_ms, 4),
            unit="ms_cpu",
            provenance="measured",
            note=(f"turns_saved x local_serve_ns x (boundary_tax-1); "
                  f"local_serve_ns={local_serve_ns}, tax~{SPAWNED_HOOK_TAX_X:.0f}x "
                  f"(BENCHMARK-AUTHORITY M3). The axis Net omits."),
        ))
    else:
        axes.append(AxisSaving(
            account="local_cpu", amount=0.0, unit="ms_cpu", provenance="measured",
            note="no local_serve_ns in report; local-CPU saving unmeasured here",
        ))

    # --- gpu_prefill: MODELED. A token proxy for the forward pass never run. The
    # prompt tokens of each saved turn are a forward pass the engine skips. This is
    # a TOKEN proxy, NOT FLOPs and NOT wall-clock  -  attention is O(L^2), so this
    # under-counts long-context saves on purpose (the safe direction).
    prefill_tok_saved = turns * prompt_tok
    axes.append(AxisSaving(
        account="gpu_prefill",
        amount=float(prefill_tok_saved),
        unit="prefill_tokens",
        provenance="modeled",
        note=("turns_saved x prompt_tokens_per_turn; a token proxy for the forward "
              "pass skipped. O(L^2) attention means this UNDER-counts long context."),
    ))

    # --- context_window: MEASURED-as-a-rate if a pollution rate is supplied, else
    # MODELED. Each saved call is a result that never occupied a window slot. With a
    # ctxmmu pollution rate we can say what fraction of results would have been paged
    # out anyway; without it we report the modeled slot count.
    if pollution_rate is not None:
        prov = "measured-rate"
        note = (f"turns_saved slots never entered the window; ctxmmu pollution_rate="
                f"{pollution_rate:.3f} supplied as the measured paged-out fraction.")
        amount = float(turns)
    else:
        prov = "modeled"
        note = ("turns_saved window slots never consumed (1 slot per saved result); "
                "supply --pollution-rate from ctxmmu for the measured paged-out fraction.")
        amount = float(turns)
    axes.append(AxisSaving(
        account="context_window", amount=amount, unit="window_slots",
        provenance=prov, note=note,
    ))

    # --- wall_clock: MODELED. The round-trips never blocked on. This is the knobbed
    # ModelTurnLatencyMs from the cost model, never a measured wall-clock.
    axes.append(AxisSaving(
        account="wall_clock",
        amount=round(latency_saved_ms, 3),
        unit="ms_wall",
        provenance="modeled",
        note=("turns_saved x cost_model.ModelTurnLatencyMs; a knobbed round-trip "
              "constant, never a measured wall-clock."),
    ))

    # The dollar saving is NOT a fifth axis  -  it is the wall_clock+prefill accounts
    # priced. We carry it only to cross-check the re-projection equals the flat Net.
    vector_dollars = dollars_saved  # by construction: same turns, same price

    binding = _pick_binding(axes, PROFILES[profile]["binding_order"], turns)

    return SavingsVector(
        turns_saved=turns, forced=forced, elision=elision,
        axes=[asdict(a) for a in axes],
        binding_account=binding, profile=profile,
        net_dollars_saved=round(dollars_saved, 6),
        vector_dollars_saved=round(vector_dollars, 6),
    )


def _pick_binding(axes, order, turns) -> str:
    """The binding account is the first in the profile's scarcity order that has a
    non-zero saving  -  i.e. the scarcest budget this run actually relieved. With no
    saving at all, nothing binds."""
    if turns == 0:
        return "none"
    have = {a.account for a in axes if a.amount > 0}
    for acct in order:
        if acct in have:
            return acct
    return order[0]


def render(v: SavingsVector) -> str:
    lines = []
    lines.append(f"savings vector (profile={v.profile})  -  {v.turns_saved} turns saved "
                 f"(forced={v.forced}, elision={v.elision})")
    lines.append(f"  binding account: {v.binding_account}")
    lines.append("")
    lines.append(f"  {'account':<16}{'saving':>16}  {'unit':<14}{'provenance':<14}")
    lines.append(f"  {'-'*16}{'-'*16:>16}  {'-'*14:<14}{'-'*14:<14}")
    for a in v.axes:
        lines.append(f"  {a['account']:<16}{a['amount']:>16,.4g}  "
                     f"{a['unit']:<14}{a['provenance']:<14}")
    lines.append("")
    lines.append("  cross-check (re-projection must equal the flat Net, not exceed it):")
    lines.append(f"    Net.dollars_saved    = {v.net_dollars_saved:.6f}")
    lines.append(f"    vector.dollars_saved = {v.vector_dollars_saved:.6f}")
    ok = abs(v.net_dollars_saved - v.vector_dollars_saved) < 1e-9
    lines.append(f"    equal: {'YES (decomposes, does not inflate)' if ok else 'NO  -  BUG'}")
    return "\n".join(lines)


def cmd_report(args) -> int:
    with open(args.path, "r", encoding="utf-8") as fh:
        report = json.load(fh)
    v = build_vector(report, profile=args.profile, pollution_rate=args.pollution_rate)
    if args.json:
        print(json.dumps(asdict(v), indent=2))
    else:
        print(render(v))
    # the invariant the doc rests on: the vector re-projects Net, never exceeds it
    if abs(v.net_dollars_saved - v.vector_dollars_saved) >= 1e-9:
        print("FAIL: vector dollar saving != Net dollar saving (inflation bug)", file=sys.stderr)
        return 1
    return 0


# A synthetic Report with the exact shape internal/turnbench emits  -  used by
# selfcheck so the tool's contract is testable with no model run. Mirrors the
# fields build_vector reads; values are illustrative, NOT a benchmark claim.
SYNTHETIC_REPORT = {
    "cost_model": {
        "prompt_tokens_per_turn": 1200,
        "completion_tokens_per_turn": 120,
        "dollars_per_mtok_in": 3.0,
        "dollars_per_mtok_out": 15.0,
        "model_turn_latency_ms": 1500.0,
    },
    "local_serve_ns": 362,  # ~M3 decide
    "net": {
        "turns_saved": 10,
        "tokens_saved": 13200,           # 10 x (1200+120)
        "dollars_saved": 0.054,          # 10 x (1200/1e6*3 + 120/1e6*15) = 10x0.0054
        "latency_saved_ms": 15000.0,     # 10 x 1500
    },
    "turn_kinds": {"forced": 6, "elision": 4},
}

# A happy-path control: zero turns saved. Every axis must be 0 (anti-inflation).
SYNTHETIC_HAPPY = {
    "cost_model": SYNTHETIC_REPORT["cost_model"],
    "local_serve_ns": 362,
    "net": {"turns_saved": 0, "tokens_saved": 0, "dollars_saved": 0.0, "latency_saved_ms": 0.0},
    "turn_kinds": {"forced": 0, "elision": 0},
}


def cmd_selfcheck(_args) -> int:
    failures = []

    # 1. The re-projection equals Net on dollars (decompose, never inflate).
    v = build_vector(SYNTHETIC_REPORT, profile="laptop")
    if abs(v.net_dollars_saved - v.vector_dollars_saved) >= 1e-9:
        failures.append("vector dollars != Net dollars (inflation)")

    # 2. Every axis saving is >= 0.
    for a in v.axes:
        if a["amount"] < 0:
            failures.append(f"negative saving on {a['account']}")

    # 3. The local_cpu axis is non-zero and MEASURED when local_serve_ns is present
    #     -  this is the whole point: the account Net omits is now surfaced.
    cpu = [a for a in v.axes if a["account"] == "local_cpu"][0]
    if cpu["provenance"] != "measured":
        failures.append("local_cpu axis is not labeled measured")
    if cpu["amount"] <= 0:
        failures.append("local_cpu saving is zero despite local_serve_ns>0 and turns>0")

    # 4. gpu_prefill and wall_clock are honestly labeled MODELED (not measured).
    for acct in ("gpu_prefill", "wall_clock"):
        ax = [a for a in v.axes if a["account"] == acct][0]
        if ax["provenance"] != "modeled":
            failures.append(f"{acct} must be labeled modeled, got {ax['provenance']}")

    # 5. The happy-path control saves EXACTLY zero on every axis (anti-inflation
    #    gate, the same one turntax-happy enforces).
    hv = build_vector(SYNTHETIC_HAPPY, profile="laptop")
    for a in hv.axes:
        if a["amount"] != 0:
            failures.append(f"happy control nonzero on {a['account']}: {a['amount']}")
    if hv.binding_account != "none":
        failures.append(f"happy control binds an account: {hv.binding_account}")

    # 6. The binding account follows the PROFILE, never hard-coded to dollars.
    #    On a laptop, context_window binds (it has a saving and is first in order);
    #    on ci-hook, local_cpu binds. Same saving, different binding.
    lap = build_vector(SYNTHETIC_REPORT, profile="laptop").binding_account
    ci = build_vector(SYNTHETIC_REPORT, profile="ci-hook").binding_account
    if lap == ci:
        failures.append(f"binding account did not vary by profile (laptop={lap}, ci={ci})")
    if lap != "context_window":
        failures.append(f"laptop profile should bind context_window, got {lap}")
    if ci != "local_cpu":
        failures.append(f"ci-hook profile should bind local_cpu, got {ci}")

    if failures:
        print("SELFCHECK FAIL:")
        for f in failures:
            print(f"  - {f}")
        return 1
    print("SELFCHECK OK  -  vector decomposes Net (does not inflate); local_cpu surfaced "
          "and measured; gpu_prefill/wall_clock honestly modeled; happy control saves 0; "
          "binding account follows the profile.")
    return 0


def main(argv=None) -> int:
    p = argparse.ArgumentParser(description=__doc__,
                                formatter_class=argparse.RawDescriptionHelpFormatter)
    sub = p.add_subparsers(dest="cmd", required=True)

    pr = sub.add_parser("report", help="decompose a turnbench Report JSON into the savings vector")
    pr.add_argument("path", help="path to a turnbench Report JSON (fak turntax --out)")
    pr.add_argument("--profile", default="laptop", choices=sorted(PROFILES),
                    help="host shape that sets the binding (scarce) account")
    pr.add_argument("--pollution-rate", type=float, default=None,
                    help="ctxmmu paged-out fraction; promotes the context axis to measured-rate")
    pr.add_argument("--json", action="store_true", help="emit the vector as JSON")
    pr.set_defaults(func=cmd_report)

    ps = sub.add_parser("selfcheck", help="anti-overclaim gate (no model run)")
    ps.set_defaults(func=cmd_selfcheck)

    args = p.parse_args(argv)
    return args.func(args)


if __name__ == "__main__":
    raise SystemExit(main())
